web/src/pages/index.tsx

```diff
@@ -70,6 +77,32 @@ const Home: NextPage = () => {
       }
   };
 
+
+  const updateCitations = (allCitations: Citation[], current: CurrentSearch) => {
+    const entryCitations = Array.from(current.citationsMap.values());
+    if (!entryCitations.some(c => !c.index)) {
+      // All of the entries citations have indexes, so there weren't any changes since the last check
+      return
+    }
+
+    // Get a mapping of all known citations, so as to reuse them if they appear again
+    const citationsMapping = Object.fromEntries(allCitations.map(c => ([c.title + c.url, c.index])));
+
+    entryCitations.forEach(
+      (c) => {
+        const hash = c.title + c.url;
+        if (!citationsMapping[hash]) {
+          c.index = allCitations.length + 1;
+          allCitations.push(c);
+        } else {
+          c.index = citationsMapping[hash];
+        }
+      }
+    )
+    setCitations(allCitations)
+    setCurrent(current)
```

Comment:
(minor) as a first impression, it looks suspicious to me to set both of these at the same event, I would expect that all citations need to be calculated less often than whatever is "current" ... but I didn't look into it in details, I'm happy if this works as intended

Reply:
Totally agreed. This whole function is really ugly :/
This function should only be called when the current text gets updated. Which is quite a few times per query, but not that much.
`current` is the results of the current query, which gets updated a lot - once it's final it gets added to the history (which are previous interactions) and current is set to null (or something like that).
`allCitations` is the list of all displayed citations.
This function needs to update both or none - if the newest text chunk doesn't contain citations (or only known ones), then neither should be updated, but if it contains a new citation, then the map of `<letter>: <citation>` in the current object needs to be updated with the new indexes, but that also implies that the list of all show citations also needs to be updated with the new citations.